packages/react/src/useActionForm.ts

```diff
@@ -0,0 +1,293 @@
+  /**
+   * Any error that occurred during initial data fetching or action submission
+   */
+  error?: ErrorWrapper | Error | null;
+  /**
+   * Function to call to submit the form
+   */
+  submit: () => Promise<void>;
+  /**
+   * The data resulting from running the action
+   */
+  actionData?: ActionFunc extends ActionFunction<GivenOptions, any, any, SchemaT, any>
+    ? ReturnType<typeof useAction<GivenOptions, SchemaT, ActionFunc, any>>[0]["data"]
+    : ReturnType<typeof useGlobalAction<any>>[0]["data"];
+};
+
+/**
+ * React hook to manage state for a form that calls a Gadget action. `useActionForm` must be passed an action function from an instance of your generated API client library, like `api.user.create`, `api.blogPost.publish` or `api.someGlobalAction`. `useActionForm` returns a `Form` object from `react-hook-form` which can be used to build great form experiences.
+ *
+ * `useActionForm` manages the Gadget API calls for the form, and handling validation errors returned by the Gadget backend. If your form is for an existing record, `useActionForm` will also fetch the record from the backend and populate the form with the existing values.
+ *
+ * `useActionForm` doesn't run the action when called. To actually submit the form, call the `submit` function on the returned `Form` object.
+ *
+ * @see https://react-hook-form.com/
+ *
+ * @param action an action function from a model manager in your application's client, like `api.user.create`
+ * @param options action options, like selecting the fields in the result
+ */
+export const useActionForm = <
+  GivenOptions extends OptionsType,
+  SchemaT,
+  ActionFunc extends ActionFunction<GivenOptions, any, any, SchemaT, any> | GlobalActionFunction<any>,
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  ExtraFormVariables = {},
+  FormContext = any,
+  FormVariables extends FieldValues = ActionFunc["variablesType"] & ExtraFormVariables
+>(
+  action: ActionFunc,
+  options?: Omit<UseFormProps<FormVariables, FormContext>, "defaultValues"> & {
+    defaultValues?: DeepPartial<FormVariables> & { [key in OmittedKey]?: any };
+    /**
+     * The record identifier to run this action on, if it already exists.
+     * Should be undefined for create actions, or a record ID (or finder) for update / etc actions
+     **/
+    findBy?: RecordIdentifier;
+    /**
+     * Which fields to select on from the record when retrieving it from the backend.
+     */
+    select?: ActionFunc extends ActionFunction<GivenOptions, any, any, SchemaT, any> ? ActionFunc["optionsType"]["select"] : never;
+    /**
+     * Which fields to send from the form's values when sending it from the backend.
+     */
+    send?: string[];
+    /**
+     * Called when the form submits
+     */
+    onSubmit?: () => void;
+    /**
+     * Called when the action completes successfully on the backend
+     */
+    onSuccess?: (actionResult: any) => void;
```

Comment:
Can we type the result? I think we type it elsewhere too

Reply:
:+1:

Reply:
done